### PR TITLE
Fix build error on some platforms

### PR DESCRIPTION
**Pull request by Jinming-Hu**
closes https://github.com/Azure/azure-sdk-for-cpp/issues/2716
closes https://github.com/Azure/azure-sdk-for-cpp/issues/2560

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

**Comment by Jinming-Hu** [comment]
@vhvb1989 @antkmsft @RickWinter Can you please help review this PR?

**Comment by antkmsft** [comment]
Do you have a repro? Can we check if this is because `uuid.hpp` includes an openssl header? The rest of the includes are in src files, and we've missed this one. The proper fix then, I think, is to move that part of a function into a non-inlineable function, which is implemented in cpp file and uses `RAND_bytes`.

**Comment by Jinming-Hu** [comment]
> @Jinming-Hu Do you have a repro? Can we check if this is because `uuid.hpp` includes an openssl header? The rest of the includes are in src files, and we've missed this one. The proper fix then, I think, is to move that part of a function into a non-inlineable function, which is implemented in cpp file and uses `RAND_bytes`.

Done, please review it again.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp [hunk 1]
```diff
@@ -44,71 +37,12 @@ namespace Azure { namespace Core {
...
-      std::random_device rd;
-
-      for (size_t i = 0; i < UuidSize; i += 4)
-      {
-        const uint32_t x = rd();
-        std::memcpy(uuid + i, &x, 4);
-      }
-#elif defined(AZ_PLATFORM_POSIX)
-      // This static cast is safe since we know Uuid size, which is a const, will always fit an int.
-      int ret = RAND_bytes(uuid, static_cast<int>(UuidSize));
```
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/uuid.hpp:95]: I was thinking of encapsulating this specific function call to a non-inline private function, but making the entire functions non-inlineable probably works too. Try to get @RickWinter to review, he wrote this.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp [hunk 2]
```diff
@@ -11,14 +11,7 @@
 #include "azure/core/platform.hpp"
 
 #include <cstring>
```
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/uuid.hpp:13]: With this header as is, with all the function bodies moved to cpp, I don't think `#niclude <cstring>` (not `<string>`) is needed anymore, so should be moved to cpp file. `<cstring>` is for `std::snprintf()`, I think.

**Review by antkmsft — APPROVED** [review]
Ideally, if you move the entirety of functions to cpp (inline => non-inline), I'd recommend to get @RickWinter 's approval as he is the author. But if Rick does not get to see these changes, I think the changes are safe. But I think that if you encapsulate just the `RAND_bytes()` call, it is less of a change (not that it is better/worse)

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 2 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:13]: `<cstdio>` is for `std::snprintf()`. `<cstring>` is for `std::memcpy`.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:95]: I don't think there's any difference. After moving the entire function body to `.cpp` file, the header file now looks more concise.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 2 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:13]: Still, if we keep it as is, with just the function signatures, I think, memcpy is not used in the header, just in the cpp.

**Comment by Jinming-Hu** [comment]
@RickWinter I moved the function body of uuid impl to `.cpp` file, so that we don't have to include openssl in header file, so as to fix #2716, #2560 these two issues.

Can you please review this PR?

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 1 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:95]: I can't disagree, it is only the perf that could be a bit different, although I don't believe it makes a difference, in this place.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:95]: If we only keep the class declaration here, I think we should move `#include <cstring>` from the header to the source.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:95]: We still need `<cstring>` because we're using `std::memcpy` in the private constructor.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:95]: Ah, I see it now, thanks for clarification! GitHub hid that line, and I didn't see it. Sure then!

**Comment by antkmsft** [comment]
I think we need a changelog entry to the "Other Changes" section, "Fixed  compilation error on POSIX platforms where openssl was not available.", something like that.

**Review by RickWinter — APPROVED** [review]

**Review by vhvb1989 — APPROVED** [review]
cool!  thank you!

**Pull request merged**